feat(perf-issues): Parametrize DB span SQL queries before hashing

mjq-sentry opened the pull request:
N+1 issue detection sometimes creates duplicate issues for what should logically be a single N+1 issue.

We have a restriction in place that N+1 performance issues are only created for spans that contain some SQL parametrization, but there are still cases of partial parametrization in the wild. When this happens inside the "N" queries, we don't create an N+1 issue when we should. When this happens inside the query that precedes the N+1, each instance gets its own fingerprint, leading to duplicate issues.

Previously our span hashing applied only very light normalization to database queries, collapsing lists of IN parameters into a single placeholder. Now, additionally apply more aggressive normalization: parametrize numberic, boolean, and single-quoted string literals in queries before generating the hash.

The regular expressions were sourced from OpenTelemetry's Apache 2.0-licensed [instrumentation for ruby's pg](https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/179f145feab24360209421b5b46deae21669e5f2/instrumentation/pg/lib/opentelemetry/instrumentation/pg/constants.rb#L69) and [mysql2 libraries](https://github.com/open-telemetry/opentelemetry-ruby-contrib/blob/ed4eec3320cc35079191416ef0cb6268fe51be60/instrumentation/mysql2/lib/opentelemetry/instrumentation/mysql2/patches/client.rb#L33), which offer similar parametrization (they call it obfuscation).

Because we don't know what flavor of SQL a particular span uses, we have to be conservative about what we parametrize. For example, we can't safely parametrize double-quoted strings even though they can be used for string literals in MySQL, because in PostgreSQL they're used to quote identifiers (table/column names, etc). The literals implemented here should be safe across all SQL flavors we'll come across.

Even though we don't parametrize every possible literal, even an incremental change towards better parametrization is good: more good N+1s, fewer dupes.

This will improve both N+1 detection (detection N+1s we previously couldn't and creating fewer duplications) and also Suspect Spans, which is driven by the same span hashes (equivalent queries will be grouped together correctly).

Note that this new hashing won't be applied to any actual transactions until the default grouping config ID is updated. We will first collect metrics for performance and hashing differences to verify the behaviour, and then update the default config if everything looks good.

mjq committed:
feat(perf-issues): Parametrize DB span SQL queries before hashing

N+1 issue detection sometimes creates duplicate issues for what should
logically be a single N+1 issue.

We have a restriction in place that N+1 performance issues are only created for
spans that contain some SQL parametrization, but there are still cases of
partial parametrization in the wild. When this happens inside the "N" queries,
we don't create an N+1 issue when we should. When this happens inside the query
that precedes the N+1, each instance gets its own fingerprint, leading to
duplicate issues.

Previously our span hashing applied only very light normalization to database
queries, collapsing lists of IN parameters into a single placeholder. Now,
additionally apply more aggressive normalization: parametrize numberic,
boolean, and single-quoted string literals in queries before generating the
hash.

Because we don't know what flavor of SQL a particular span uses, we have to be
conservative about what we parametrize. For example, we can't safely
parametrize double-quoted strings even though they can be used for string
literals in MySQL, because in PostgreSQL they're used to quote identifiers
(table/column names, etc). The literals implemented here should be safe across
all SQL flavors we'll come across.

Even though we don't parametrize every possible literal, even an incremental
change towards better parametrization is good: more good N+1s, fewer dupes.

This will improve both N+1 detection (detection N+1s we previously couldn't and
creating fewer duplications) and also Suspect Spans, which is driven by the
same span hashes (equivalent queries will be grouped together correctly).

Note that this new hashing won't be applied to any actual transactions until
the default grouping config ID is updated. We will first collect metrics for
performance and hashing differences to verify the behaviour, and then update
the default config if everything looks good.

k-fish commented on src/sentry/spans/grouping/strategy/base.py:
```diff
@@ -135,6 +135,32 @@ def loose_normalized_db_span_in_condition_strategy(span: Span) -> Optional[Seque
+
+
+DB_PARAMETRIZATION_PATTERN = re.compile(
+    join_regexes(
+        [
+            r"'(?:[^']|'')*?(?:\\'.*|'(?!'))",  # single-quoted strings
+            r"-?\b(?:[0-9]+\.)?[0-9]+(?:[eE][+-]?[0-9]+)?\b",  # numbers
+            r"\b(?:true|false)\b",  # booleans
+        ]
+    )
```
iirc it's been a bit, and I think our input is bounded enough and it looks like just capture groups so that it won't be an issue, but in general regex + lots of data can cause performance overhead, especially since for a single event we can run up into the hundreds or thousands of sql statements to compare this against. Even though this is in 'dry-run' mode the performance impact still applies to production, I'd maybe just check quickly locally if you have one of those long non-truncated sql statements.

mjq-sentry replied:
Sure, I'll find some SQL-heavy events to benchmark against locally :+1: Good call, thanks!

k-fish replied:
yep, was just a concern since it's one of those possible unintentional leaks even though this is supposed to be 'dry-run' so brought it having hit regex being troublesome in the past.

k-fish reviewed (approved):
Minus comment about perf overhead which I'll leave to you to confirm, looks good! 🚢

gggritso commented on src/sentry/spans/grouping/strategy/base.py:
```diff
@@ -135,6 +135,32 @@ def loose_normalized_db_span_in_condition_strategy(span: Span) -> Optional[Seque
+
+@span_op(["db", "db.query", "db.sql.query", "db.sql.active_record"])
+def parametrize_db_span_strategy(span: Span) -> Optional[Sequence[str]]:
+    """Something"""
```
😂

mjq-sentry replied:
...missed that one 😂

mjq-sentry replied:
re: negative tests and unexpected literal tests - these are there on the `parametrize_db_span_strategy` unit tests:
```
("SELECT * FROM table123 WHERE id = %s", None),
("SELECT * FROM table_true WHERE deleted = %s", None),
```
and I used `LIMIT` instead of `OFFSET` but
```
("SELECT * FROM table LIMIT 1", ["SELECT * FROM table LIMIT %s"]),
```

Do those look sufficient or should I add more?

Good point about the loose IN tests, I'll add those to the `parametrize_db_span_strategy` tests as well.

BTW, re: stacking the loose IN strategy: I looked into it originally but I don't think it will work because, to my understanding, [we stop looking at strategies as soon as one returns a value](https://github.com/getsentry/sentry/blob/d6eb9fd46c5a7b1da23b710e30238bb5dc3f0075/src/sentry/spans/grouping/strategy/base.py#L71-L77). The end result is we can only really have one strategy per op.

gggritso replied:
Gotcha! In that case LGTM, up to you on the test coverage! Resolving

gggritso reviewed (approved):
Awesome work! One minor comment: since this is dealing with regular expressions, it might be good to beef up the test suite a bit. e.g.,

- adding negative test cases like `"SELECT * FROM project_11"` or `"SELECT * FROM true_values"` to make sure number and boolean literals don't get unnecessarily stripped
- add some cases similar to real data we've seen, like `"SELECT * FROM some_table OFFSET 15"` that have literals in unexpected places
- since the new strategy combines parameterization _and_ the loose IN strategy, I think you should either add all the unit tests from the loose IN strategy _or_ stack those strategies in the config

mjq committed:
Address review comments

- Added proper docstring to strategy
- Added more negative tests
- Added all tests from the loose IN detector - and found a bug as a result.
  Flipped the order of the regexes (first LOOSE_IN, then PARAMETRIZE) to
  resolve the issue.

mjq-sentry committed:
Merge branch 'master' into mjq/db-span-parameterization
